### DISCUSSION TITLE
Having trouble wrapping query_map into an iterator

### DISCUSSION THREAD

**Discussion by sartak** [rusqlite/rusqlite#1198] [General]
Hi Rusqlite devs!

I'm trying to wrap Rusqlite's `query_map` API to return an iterator of structs. But I'm hitting trouble with lifetimes and can't seem to make progress. This is the sort of code I'm trying to write, but it complains that "cannot return value referencing local variable `stmt`"

```rust
use rusqlite::{Connection, Result};

struct Person {
    id: i32,
    name: String,
}

fn get_people(conn: Connection) -> Result<Box<dyn Iterator<Item = Result<Person>>>> {
    let mut stmt = conn.prepare("SELECT rowid, name FROM people")?;
    let results = stmt.query_map([], |row| {
        Ok(Person {
            id: row.get(0)?,
            name: row.get(1)?,
        })
    })?;

    Ok(Box::new(results))
}

fn main() {
    let conn = Connection::open("test.sqlite").unwrap();
    let people = get_people(conn).unwrap();

    for person in people {
        let person = person.unwrap();
        println!("person {}: {}", person.id, person.name);
    }
}
```

I also tried creating a new struct that packages together `stmt` and `results` and implements `Iterator`, but that had similar issues where I "cannot move out of `stmt` because it is borrowed" (by the call to `query_map`).

All this means I'm having trouble doing better than loading all results into memory and returning a `Vec<Person>`, which is not ideal. Do you have any advice for what I can do here? Is there a more appropriate Rusqlite API for this besides `query_map`? (I also confess to be a Rust newbie so it's entirely possible I'm just missing some intermediate language feature!)

Thanks very much :)

**Comment by gwenn** [comment]
[collect()](https://docs.rs/rusqlite/latest/rusqlite/struct.MappedRows.html) ?

**Comment by adumbidiot** [comment]
I can see 2 issues: 
1. You pass the connection to the `get_people` function by value. When the connection is dropped inside that function, the db should be inaccessible. As a result, the db can only be used inside that function if you pass by value. I would suggest passing by reference here.
2. You are trying to return the `MappedRows` iter from a function that internally creates a statement. Rust doesn't easily support self-referential structs, like the one you would need to return that iter from the function. There are a few solutions. The easiest would be to just create the reference outside the `get_people` function and pass it by reference to that function. The returned iterator would not need to be a self-referential struct in that case. The second option is creating a self-referential struct. Doing so is very difficult and easy to mess up, so using an external crate like `ouroboros` is a good idea:
```rust
struct Person {
    id: i32,
    name: String,
}

fn get_people(conn: &Connection) -> Result<impl Iterator<Item = Result<Person>> + '_> {
    let statement = conn.prepare("SELECT rowid, name FROM people")?;

    Ok(PeopleIterTryBuilder {
        statement,
        iter_builder: |statement| {
            statement.query_map([], |row| {
                Ok(Person {
                    id: row.get(0)?,
                    name: row.get(1)?,
                })
            })
        },
    }
    .try_build()?)
}

#[ouroboros::self_referencing]
struct PeopleIter<'a, F> {
    statement: rusqlite::Statement<'a>,

    #[borrows(mut statement)]
    #[covariant]
    iter: rusqlite::MappedRows<'this, F>,
}

impl<'a, F> Iterator for PeopleIter<'a, F>
where
    F: FnMut(&rusqlite::Row<'_>) -> Result<Person>,
{
    type Item = Result<Person>;

    fn next(&mut self) -> Option<Self::Item> {
        self.with_iter_mut(|iter| iter.next())
    }
}
```

However, doing all this just to avoid returning a `Vec<Person>` probably isn't worth it. If you want to use multiple connections, a shared lock will be held on the database while reading, preventing writers from accessing the database, unless you use something like WAL. However, this only partially alleviates the issue as it can still prevent checkpointing. As a result, it is better to access the database as quickly as possible. If you are using a single connection, you should consider whether all this complexity is worth the time you would save by just returning a `Vec`, as you would need to implement a custom self-referential struct for each type you want to access from the database, or create statements outside of these functions before passing them in.

**Comment by sartak** [comment]
Thanks so much @adumbidiot!!
